woocommerce/woocommerce-gutenberg-products-block
# Add eslint to ensure correct usage of feature flag

This PR builds on #1631, it adds an eslint rule `woocommerce/feature-flag` that ensures the flag is correctly being used, it also add auto fixer, and this is actually the most exciting part for me 😆

This rule builds heavily (read: copy paste) from Gutenberg rule, it’s adapted to a set of defined strings instead of numbers and includes auto fixer (that I might push later to GB as well).

----
### 1. Whitelisted Flags only
> The `WOOCOMMERCE_BLOCKS_PHASE` constant should only be used in a strict equality comparison with a predefined flag of (experimental, stable).

**auto fixable** to `experimental`
```js
// good
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'stable' ) {

// bad
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'plugin' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'core' ) {
```

### 2. Strict equal only
> The `WOOCOMMERCE_BLOCKS_PHASE` comparison should only be a strict equal `===`, if you need `!==` try switching the flag.

**auto fixable** to `===`
```js
// good
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'stable' ) {

// bad
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE !== 'experimental' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE > 'experimental' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE < 'experimental' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE == 'experimental' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE >= 'experimental' ) {
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE <= 'experimental' ) {
```

### 3. Correct access to the flag
> The `WOOCOMMERCE_BLOCKS_PHASE` constant should be accessed using `process.env.WOOCOMMERCE_BLOCKS_PHASE`.

**auto fixable** to `process.env.WOOCOMMERCE_BLOCKS_PHASE`
```js
// good
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {

// bad
if ( WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
if ( env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
if ( process.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {
```

### 4. If statement only (ternary can’t be treeshaken)
> The `WOOCOMMERCE_BLOCKS_PHASE` constant should only be used as part of the condition in an if statement or ternary expression.

```js
// good
if ( process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental' ) {

// bad
const isFeatureActive = process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental';
```

## How to test
If you have [ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) setup in your vscode editor:
1. after pulling this PR, reload your editor (command + shift + p > reload window).
2. write any bad expression as you see above
3. see the red lines
4. if you have autofix on save, try saving an autofixable code, it should fix it for you.

if not, run `npm run lint:js` with wrong code

```
/assets/js/blocks/cart-checkout/checkout/index.js
  57:6  error  The `WOOCOMMERCE_BLOCKS_PHASE` constant should be accessed using `process.env.WOOCOMMERCE_BLOCKS_PHASE`  woocommerce/feature-flag
```

you can also run tests directly

```
npm run test ./bin/eslint-plugin-woocommerce/rules/__tests__/feature-flag.js
```

## Screenshots

![image](https://user-images.githubusercontent.com/6165348/73592001-bace9b00-44f5-11ea-9de8-afcf0e1c1cd1.png)
![image](https://user-images.githubusercontent.com/6165348/73592010-d174f200-44f5-11ea-836d-b268e0212244.png)
![image](https://user-images.githubusercontent.com/6165348/73592024-0123fa00-44f6-11ea-9c4a-25c943878c79.png)
